File: week1_baseline/ruby/12_context/lib/boukensha/registry.rb
require_relative "errors"
require "opentelemetry/sdk"

module Boukensha
  class Registry
    def initialize(context)
      @context = context
    end

    def tool(name, description:, parameters: {}, &block)
      tool = Tool.new(name.to_s, description, parameters, block)
      @context.register_tool(tool)
      tool
    end

    def dispatch(name, args = {}, tool_call_id: nil)
      Boukensha.tracer.in_span('tool.dispatch') do |span|
        span.set_attribute('tool.name', name.to_s)
        span.set_attribute('tool.args', args.to_s)
        span.set_attribute('gen_ai.operation.name', 'execute_tool')
        span.set_attribute('gen_ai.tool.name', name.to_s)
        span.set_attribute('gen_ai.tool.call.arguments', args.to_s)
        span.set_attribute('gen_ai.tool.call.id', tool_call_id.to_s) if tool_call_id

        tool = @context.tools[name.to_s]
        raise UnknownToolError, "No tool registered as '#{name}'" unless tool

        result = tool.block.call(**args.transform_keys(&:to_sym))
        span.set_attribute('tool.result', result.to_s[0..200])
        span.set_attribute('gen_ai.tool.call.result', result.to_s[0..200])
        result
        # No explicit rescue here: Tracer#in_span already records the
        # exception and sets error status on this span for anything that
        # escapes the block above (unknown tool or a failing tool.block).
      end
    end
  end
end

File: week1_baseline/python/12_context/boukensha/registry.py
from .tool import Tool
from .errors import UnknownToolError
from .telemetry import tracer


class Registry:
    def __init__(self, context):
        self.context = context

    def tool(self, name, description, parameters=None, block=None):
        tool = Tool(str(name), description, parameters or {}, block)
        self.context.register_tool(tool)
        return tool

    def dispatch(self, name, args=None, tool_call_id=None):
        args = args or {}
        with tracer().start_as_current_span("tool.dispatch") as span:
            span.set_attribute("tool.name", str(name))
            span.set_attribute("tool.args", str(args))
            span.set_attribute("gen_ai.operation.name", "execute_tool")
            span.set_attribute("gen_ai.tool.name", str(name))
            span.set_attribute("gen_ai.tool.call.arguments", str(args))
            if tool_call_id:
                span.set_attribute("gen_ai.tool.call.id", str(tool_call_id))

            tool = self.context.tools.get(str(name))
            if tool is None:
                raise UnknownToolError(f"No tool registered as '{name}'")

            result = tool.block(**args)
            span.set_attribute("tool.result", str(result)[:200])
            span.set_attribute("gen_ai.tool.call.result", str(result)[:200])
            return result
            # No explicit except here: start_as_current_span already records
            # the exception and sets error status on this span for anything
            # that escapes the block above (unknown tool or a failing
            # tool.block).
